keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,662 @@
+</pre>
+
+On k8s, the pod and runtime class yamls would look as follows:
+
+<pre>
+apiVersion: node.k8s.io/v1
+kind: RuntimeClass
+metadata:
+  name: runhcs-wcow-process
+handler: runhcs-wcow-process 
+</pre>
+
+Corresponding pod definition would look as follows:
+
+<pre>
+apiVersion: v1 
+kind: Pod 
+metadata: 
+  name: mypod 
+spec:
+  runtimeClassName: runhcs-wcow-runhcs
+  containers: 
+  - name: busybox 
+    image: docker.io/library/busybox:1.35.0-musl 
+</pre>
+
+This document details the changes needed to support pulling images per runtime class.
+Adding this support would enable us to have different versions of the same image on a given node and could be really beneficial in virtualization based runtimes. It will also be very helpful to support future projects to support snapshotter per runtimeclass. 
+For windows in particular, different versions of a multiarch image could be pulled based on hyperV and process isolation and we could remove the necessity for exact host and guest OS version matches for hyperV isolation.
+
+This KEP attempts to goes over the details of the most significant changes needed in CRI, kubelet and tools like ctr and crictl to support this feature. Container runtimes like containerd and CRI-O are responsible for making necessary changes in their runtime to support this feature. Backwards compatibility will always be maintained to ensure existing behavior in container runtimes do not break.
+
+## Motivation
+
+Currently, the windows platform matcher looks for an exact OS version match between the host and guest while pulling images. This behavior is not accurate for hyperV isolation as the UVM/sandbox OS version could be different from the host OS version.  
+
+For example, a user might want to pull an image like python (a manifest list) on a WS2022 host with an intent of running it in on a WS2019 hyperV isolated UVM. But unfortunately, the current windows platform matcher behavior only forces the WS2022 version of the python image to be pulled (due to the exact version match between host and guest). If a manifest list does not have a version that matches the host OS version, then the image pull fails.
+Extending image pulls to specify an optional runtime class would greatly help to extend windows platform matcher behavior to pull different manifests from the image index depending on what runtime class is used.
+For example, we could have one runtime class pull one version of python image while another runtime class could be used to pull a completely different version of the same image.
+Additionally, having support to pull images per runtime class might also help projects like supporting different snapshotters with different runtimes as well.
+
+### Goals
+
+This KEP describes the detailed design and API changes needed in kubelet and CRI to support image pull with optional runtime class option.
+Goal is to keep the functionality same as it is today if user does not use the runtime class option with image pull.
```

Comment:
** if we add platform to containerd runtime class....and change the platform matcher in containerd.. then the user/admin can still pick on a "node" basis what the behavior will be for said default/existing runtime classes.

Reply:
Hmm not sure I understand the concern here :) Are you suggesting this be added as smth the KEP could achieve?

Reply:
not really a concern.. this is a more a mitigation/statement regarding what could be done "if user does not use the runtime class option with image pull" in regards to the goal stated here "to keep the functionality same as it is today."